src/rgw/rgw_auth_s3.cc

```diff
@@ -1225,6 +1227,11 @@ size_t AWSv4ComplMulti::recv_body(char* const buf, const size_t buf_max)
     to_extract -= received;
   }
 
+  if(buf_pos < buf_max && received > 0) {
+    dout(30) << "AWSv4ComplMulti: get next chunk, total buffer:" << buf_max <<", filled=" << buf_pos << dendl;
+    goto again;
```

Comment:
thanks @VVoidV, i definitely agree that we should loop until we've filled the input buffer. the use of `goto` here does simplify the diff, but i'd prefer to use a real loop construct instead

what do you think about renaming this `AWSv4ComplMulti::recv_body()` function to `recv_chunk()` and adding a new `recv_body()` that loops over calls to `recv_chunk()`? i think the only necessary addition to `recv_chunk()` is a `bool& eof` argument that gets set in the `received == 0` cases

Reply:
Thanks @cbodley, the loop version is done.  Please take a look.